src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -383,7 +383,96 @@ protected void InitializeReferences()
             RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
             try
             {
-                return runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                bool skipScan = false;
+                runner.InitializeTimeout(internalMatchTimeout);
+                ReadOnlySpan<char> span = input.AsSpan(beginning, length);
+                runner.InitializeForScan(this, span, 0, span.Length, startat - beginning, quick);
+                runner.InitializeForGo();
+
+                int stoppos = RightToLeft ? 0 : span.Length;
+
+                // If previous match was empty or failed, advance by one before matching.
+                if (prevlen == 0)
+                {
+                    if (runner.runtextstart == stoppos)
+                    {
+                        skipScan = true;
+                        runner.runmatch = System.Text.RegularExpressions.Match.Empty;
+                    }
+
+                    runner.runtextpos += RightToLeft ? -1 : 1;
+                }
+
+                if (!skipScan)
+                {
+                    runner.Scan(span);
+
+                    // if we got a match, set runmatch to null if quick is true
+                    if (runner.runmatch!._matchcount[0] > 0)
+                    {
+                        if (quick)
+                        {
+                            runner.runmatch = null;
+                        }
+                    }
```

Comment:
As noted, we really need to not do this.  This is a ship stopper.

Reply:
Yeah, this little implementation detail is there as part of a copy paste from code that used to be in individual implementations of Scan on the engine, to code that now lives in Run(), and it was originally meant to serve for IsMatch() purposes where a null match means "The input is a match for pattern". Now that all of that lives in Run, we don't have to reset runmatch, only to return null from Run() (which we couldn't do from the new Scan since it returns void)